[quant] Implement PTQ for APoT FakeQuant

asl3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81040

### Summary:
This PR implements PTQ for APoT FakeQuant. It runs models (Resnet-18 pre-trained model, ImageNet dataset) to compare accuracy metrics for different qconfig settings of uniform vs. APoT quantized activation and weight. 

According to the collected accuracy stats, model #2 (uniform activation and APoT weight) appears to have a slight improvement in accuracy compared to model #1 (uniform activation and uniform weight) for 8-bit and significant improvement for 4-bit (see "Accuracy Stats" section below).

### Test Plan:
Run models with: `python test/quantization/core/experimental/fx_graph_mode_apot.py`

### Accuracy Stats: 
8-bit (Uniform int8, APoT b = 8 k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.43% (Top-1), 85.62% (Top-5)

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.51% (Top-1), 85.78% (Top-5)

**Model #3:** APoT activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.32% (Top-1), 85.78% (Top-5)

4-bit (Uniform int4, APoT b = 4 k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 45.63% (Top-1), 71.96% (Top-5)

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.24% (Top-1), 85.56% (Top-5)

**Model #3:** APoT activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 45.40% (Top-1), 76.21% (Top-5)

**Full Precision model (FX Graph Mode quantized)**
Evaluation accuracy on test dataset: 69.76% (Top-1), 89.08% (Top-5)

**Eager mode quantized model**
Evaluation accuracy on test dataset: 69.49% (Top-1), 88.90% (Top-5)

asl3 committed:
[quant] Implement QAT for APoT FakeQuant

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement QAT for APoT FakeQuant"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement QAT for APoT FakeQuant"


### Summary:
This PR implements FX Graph Mode QAT for APoT FakeQuant.

### Test Plan:
Run models with: `python test/quantization/core/experimental/fx_graph_mode_apot.py`

Accuracy Stats:
Model #1: uniform activation, uniform weight (FX Graph Mode quantized)
Size of model (MB): 46.801265
Evaluation accuracy on test dataset: 69.76%, 89.08%

Model #2: uniform activation, APoT weight (FX Graph Mode quantized)
Size of model (MB): 46.820369
Evaluation accuracy on test dataset: 69.00%, 88.66%

Model #3: APoT activation and weight (FX Graph Mode quantized)
Size of model (MB): 46.801431
Evaluation accuracy on test dataset: 69.76%, 89.08%

Eager mode quantized model Resnet18
Size of model (MB): 11.839989
Evaluation accuracy on test dataset: 69.49%, 88.90%

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement QAT for APoT FakeQuant"


### Summary:
This PR implements FX Graph Mode QAT for APoT FakeQuant.

### Test Plan:
Run models with: `python test/quantization/core/experimental/fx_graph_mode_apot.py`

### Accuracy Stats: 
Uniform: int8
APoT: 8-bit (b = 8, k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 69.54%, 88.99%

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 69.76%, 89.08%

**Model #3:** APoT activation and weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 69.54%, 89.04%

**Model #4:** Eager mode quantized model Resnet18
Evaluation accuracy on test dataset: 69.49%, 88.90%

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement QAT for APoT FakeQuant"


### Summary:
This PR implements FX Graph Mode QAT for APoT FakeQuant.

### Test Plan:
Run models with: `python test/quantization/core/experimental/fx_graph_mode_apot.py`

### Accuracy Stats: 
Uniform: int8
APoT: 8-bit (b = 8, k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 69.54%, 88.99%

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 69.76%, 89.08%

**Model #3:** APoT activation and weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 69.54%, 89.04%

**Model #4:** Eager mode quantized model Resnet18
Evaluation accuracy on test dataset: 69.49%, 88.90%

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement QAT for APoT FakeQuant"


### Summary:
This PR implements FX Graph Mode QAT for APoT FakeQuant.

### Test Plan:
Run models with: `python test/quantization/core/experimental/fx_graph_mode_apot.py`

### Accuracy Stats: 
Uniform: int8
APoT: 8-bit (b = 8, k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 69.54%, 88.99%

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 69.76%, 89.08%

**Model #3:** APoT activation and weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 69.54%, 89.04%

**Model #4:** Eager mode quantized model Resnet18
Evaluation accuracy on test dataset: 69.49%, 88.90%

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement QAT for APoT FakeQuant"


### Summary:
This PR implements FX Graph Mode QAT for APoT FakeQuant.

### Test Plan:
Run models with: `python test/quantization/core/experimental/fx_graph_mode_apot.py`

### Accuracy Stats: 
Uniform: int8
APoT: 8-bit (b = 8, k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 69.54%, 88.99%

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 69.76%, 89.08%

**Model #3:** APoT activation and weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 69.54%, 89.04%

**Model #4:** Eager mode quantized model Resnet18
Evaluation accuracy on test dataset: 69.49%, 88.90%

[ghstack-poisoned]

dzdang commented:
this PR is for PTQ and not QAT?

asl3 committed:
Update on "[quant] Implement PTQ for APoT FakeQuant"


### Summary:
This PR implements FX Graph Mode QAT for APoT FakeQuant.

### Test Plan:
Run models with: `python test/quantization/core/experimental/fx_graph_mode_apot.py`

### Accuracy Stats: 
Uniform: int8
APoT: 8-bit (b = 8, k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 69.54%, 88.99%

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 69.76%, 89.08%

**Model #3:** APoT activation and weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 69.54%, 89.04%

**Model #4:** Eager mode quantized model Resnet18
Evaluation accuracy on test dataset: 69.49%, 88.90%

[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/experimental/fx_graph_mode_apot.py:
```diff
@@ -0,0 +1,272 @@
+b=4, k=2
+"""
+
+qconfig1 = get_uniform_qconfig_4bit()
+qconfig_dict1 = {"object_type": [(torch.nn.Linear, qconfig1)]}
+prepared_model1 = prepare_fx(float_model, qconfig_dict1)  # fuse modules and insert observers
+calibrate(prepared_model1, data_loader_test)  # run calibration on sample data
+quantized_model1 = convert_fx(prepared_model1)  # convert the calibrated model to a quantized model
+
+top1, top5 = evaluate(quantized_model1, criterion, data_loader_test)
+print("Model #1 Evaluation accuracy on test dataset (b=4, k=2): %2.2f, %2.2f" % (top1.avg, top5.avg))
```
can we define a helper function for these things?

jerryzh168 commented on test/quantization/core/experimental/fx_graph_mode_apot.py:
```diff
@@ -0,0 +1,272 @@
+
+qconfig1 = get_uniform_qconfig_4bit()
+qconfig_dict1 = {"object_type": [(torch.nn.Linear, qconfig1)]}
+prepared_model1 = prepare_fx(float_model, qconfig_dict1)  # fuse modules and insert observers
```
do we need to do a deepcopy for float_model?

jerryzh168 commented on test/quantization/core/experimental/fx_graph_mode_apot.py:
```diff
@@ -0,0 +1,272 @@
+Prepare eager mode quantized model
+"""
+
+from torchvision.models.quantization.resnet import resnet18
```
would be good to put this to a separate function as well

asl3 replied:
since this is the only instance of eager mode quantization maybe i will leave outside? not sure if a helper function would simplify much

jerryzh168 commented on torch/ao/quantization/experimental/apot_utils.py:
```diff
@@ -5,52 +5,64 @@
-    levels_lst = list(levels)
-    indices_lst = list(indices)
+    if round(x, 4) in float_to_apot_cache:
+        return float_to_apot_cache[round(x, 4)]
```
wait, doesn't this cache depend on the quantization parameters? i.e. if we change `levels`, 'indicies', 'alpha' this will break right?

asl3 replied:
oh yes, i was running each instance of quantization parameters independently so it wouldn't be an issue but that would be an issue if they were run together -- will fix!

jerryzh168 commented on torch/ao/quantization/experimental/qconfig.py:
```diff
@@ -0,0 +1,77 @@
+apot_qconfig_4bit = QConfig(activation=APoTFakeQuantize.with_args(b=4, k=2, dtype=torch.quint8),
+                            weight=APoTFakeQuantize.with_args(b=4, k=2, dtype=torch.qint8))
+
+def get_uniform_qconfig_8bit():
```
feels like no need to define these functions, we can just use the variable directly

jerryzh168 commented on torch/ao/quantization/experimental/apot_utils.py:
```diff
@@ -5,52 +5,64 @@
-
-    min_delta = math.inf
-    best_fp = 0.0
+def float_to_reduced_precision(x, levels):
```
nit: it would clearer if we name this "quant_dequant_apot" I think

jerryzh168 commented on torch/ao/quantization/experimental/fake_quantize.py:
```diff
@@ -1,34 +1,45 @@
         super().__init__()
-        self.activation_post_process = APoTObserver(**observer_kwargs)
+        self.activation_post_process = observer(**observer_kwargs)
+        dtype = observer_kwargs.get("dtype", torch.quint8)
+        if hasattr(observer, "p"):
+            # If observer is _PartialWrapper, dtype can be stored in
+            # observer.p.keywords["dtype"]
+            dtype = getattr(getattr(observer, "p", {}), "keywords", {}).get(
+                "dtype", dtype)
```
I think this can be removed?

jerryzh168 commented on torch/ao/quantization/experimental/fake_quantize.py:
```diff
@@ -1,34 +1,45 @@
-            X = dequantize_APoT(X)
+            X = quant_dequant_APoT(X, self.alpha, self.gamma, self.quantization_levels, self.level_indices)
 
+        print("finished forward")
```
remove

asl3 committed:
Update on "[quant] Implement PTQ for APoT FakeQuant"


### Summary:
This PR implements FX Graph Mode QAT for APoT FakeQuant.

### Test Plan:
Run models with: `python test/quantization/core/experimental/fx_graph_mode_apot.py`

### Accuracy Stats: 
8-bit (Uniform int8, APoT b = 8 k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.43%, 85.62%

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.51%, 85.78%

**Model #3:** APoT activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.32%, 85.78%

4-bit (Uniform int4, APoT b = 8 k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 45.63, 71.96

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.24%, 85.56%

**Model #3:** APoT activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 45.40%, 76.21%

**Full Precision model (FX Graph Mode quantized)**
Evaluation accuracy on test dataset: 69.76%, 89.08%

**Eager mode quantized model**
Evaluation accuracy on test dataset: 69.49%, 88.90%

[ghstack-poisoned]

dzdang commented:
PR title mismatch with summary

jerryzh168 commented on torch/ao/quantization/fake_quantize.py:
```diff
@@ -328,6 +328,8 @@ def forward(self, X: torch.Tensor) -> torch.Tensor:
 Default fake_quant for activations.
 """
 
+default_symmetric_fake_quant = FakeQuantize.with_args(observer=MovingAverageMinMaxObserver, qscheme=torch.per_tensor_symmetric)
```
would be nice if we specify dtype as well I think, just to be more explicit

jerryzh168 reviewed (approved):
looks good, thanks for addressing the comments

asl3 committed:
Update on "[quant] Implement PTQ for APoT FakeQuant"


### Summary:
This PR implements PTQ for APoT FakeQuant.

### Test Plan:
Run models with: `python test/quantization/core/experimental/fx_graph_mode_apot.py`

### Accuracy Stats: 
8-bit (Uniform int8, APoT b = 8 k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.43%, 85.62%

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.51%, 85.78%

**Model #3:** APoT activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.32%, 85.78%

4-bit (Uniform int4, APoT b = 8 k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 45.63%, 71.96%

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.24%, 85.56%

**Model #3:** APoT activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 45.40%, 76.21%

**Full Precision model (FX Graph Mode quantized)**
Evaluation accuracy on test dataset: 69.76%, 89.08%

**Eager mode quantized model**
Evaluation accuracy on test dataset: 69.49%, 88.90%

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement PTQ for APoT FakeQuant"


### Summary:
This PR implements PTQ for APoT FakeQuant.

### Test Plan:
Run models with: `python test/quantization/core/experimental/fx_graph_mode_apot.py`

### Accuracy Stats: 
8-bit (Uniform int8, APoT b = 8 k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.43%, 85.62%

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.51%, 85.78%

**Model #3:** APoT activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.32%, 85.78%

4-bit (Uniform int4, APoT b = 8 k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 45.63%, 71.96%

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.24%, 85.56%

**Model #3:** APoT activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 45.40%, 76.21%

**Full Precision model (FX Graph Mode quantized)**
Evaluation accuracy on test dataset: 69.76%, 89.08%

**Eager mode quantized model**
Evaluation accuracy on test dataset: 69.49%, 88.90%

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement PTQ for APoT FakeQuant"


### Summary:
This PR implements PTQ for APoT FakeQuant.

### Test Plan:
Run models with: `python test/quantization/core/experimental/fx_graph_mode_apot.py`

### Accuracy Stats: 
8-bit (Uniform int8, APoT b = 8 k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.43%, 85.62%

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.51%, 85.78%

**Model #3:** APoT activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.32%, 85.78%

4-bit (Uniform int4, APoT b = 8 k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 45.63%, 71.96%

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.24%, 85.56%

**Model #3:** APoT activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 45.40%, 76.21%

**Full Precision model (FX Graph Mode quantized)**
Evaluation accuracy on test dataset: 69.76%, 89.08%

**Eager mode quantized model**
Evaluation accuracy on test dataset: 69.49%, 88.90%

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g

asl3 committed:
Update on "[quant] Implement PTQ for APoT FakeQuant"


### Summary:
This PR implements PTQ for APoT FakeQuant. It runs models (Resnet-18 pre-trained model, ImageNet dataset) to compare accuracy metrics for different qconfig settings of uniform vs. APoT quantized activation and weight. 

According to the collected accuracy stats, model #2 (uniform activation and APoT weight) appears to have a slight improvement in accuracy compared to model #1 (uniform activation and uniform weight) for 8-bit and significant improvement for 4-bit (see "Accuracy Stats" section below).

### Test Plan:
Run models with: `python test/quantization/core/experimental/fx_graph_mode_apot.py`

### Accuracy Stats: 
8-bit (Uniform int8, APoT b = 8 k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.43% (Top-1), 85.62% (Top-5)

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.51% (Top-1), 85.78% (Top-5)

**Model #3:** APoT activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.32% (Top-1), 85.78% (Top-5)

4-bit (Uniform int4, APoT b = 4 k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 45.63% (Top-1), 71.96% (Top-5)

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.24% (Top-1), 85.56% (Top-5)

**Model #3:** APoT activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 45.40% (Top-1), 76.21% (Top-5)

**Full Precision model (FX Graph Mode quantized)**
Evaluation accuracy on test dataset: 69.76% (Top-1), 89.08% (Top-5)

**Eager mode quantized model**
Evaluation accuracy on test dataset: 69.49% (Top-1), 88.90% (Top-5)

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement PTQ for APoT FakeQuant"


### Summary:
This PR implements PTQ for APoT FakeQuant. It runs models (Resnet-18 pre-trained model, ImageNet dataset) to compare accuracy metrics for different qconfig settings of uniform vs. APoT quantized activation and weight. 

According to the collected accuracy stats, model #2 (uniform activation and APoT weight) appears to have a slight improvement in accuracy compared to model #1 (uniform activation and uniform weight) for 8-bit and significant improvement for 4-bit (see "Accuracy Stats" section below).

### Test Plan:
Run models with: `python test/quantization/core/experimental/fx_graph_mode_apot.py`

### Accuracy Stats: 
8-bit (Uniform int8, APoT b = 8 k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.43% (Top-1), 85.62% (Top-5)

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.51% (Top-1), 85.78% (Top-5)

**Model #3:** APoT activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.32% (Top-1), 85.78% (Top-5)

4-bit (Uniform int4, APoT b = 4 k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 45.63% (Top-1), 71.96% (Top-5)

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.24% (Top-1), 85.56% (Top-5)

**Model #3:** APoT activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 45.40% (Top-1), 76.21% (Top-5)

**Full Precision model (FX Graph Mode quantized)**
Evaluation accuracy on test dataset: 69.76% (Top-1), 89.08% (Top-5)

**Eager mode quantized model**
Evaluation accuracy on test dataset: 69.49% (Top-1), 88.90% (Top-5)

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g

asl3 committed:
Update on "[quant] Implement PTQ for APoT FakeQuant"


### Summary:
This PR implements PTQ for APoT FakeQuant. It runs models (Resnet-18 pre-trained model, ImageNet dataset) to compare accuracy metrics for different qconfig settings of uniform vs. APoT quantized activation and weight. 

According to the collected accuracy stats, model #2 (uniform activation and APoT weight) appears to have a slight improvement in accuracy compared to model #1 (uniform activation and uniform weight) for 8-bit and significant improvement for 4-bit (see "Accuracy Stats" section below).

### Test Plan:
Run models with: `python test/quantization/core/experimental/fx_graph_mode_apot.py`

### Accuracy Stats: 
8-bit (Uniform int8, APoT b = 8 k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.43% (Top-1), 85.62% (Top-5)

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.51% (Top-1), 85.78% (Top-5)

**Model #3:** APoT activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.32% (Top-1), 85.78% (Top-5)

4-bit (Uniform int4, APoT b = 4 k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 45.63% (Top-1), 71.96% (Top-5)

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.24% (Top-1), 85.56% (Top-5)

**Model #3:** APoT activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 45.40% (Top-1), 76.21% (Top-5)

**Full Precision model (FX Graph Mode quantized)**
Evaluation accuracy on test dataset: 69.76% (Top-1), 89.08% (Top-5)

**Eager mode quantized model**
Evaluation accuracy on test dataset: 69.49% (Top-1), 88.90% (Top-5)

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g

asl3 committed:
Update on "[quant] Implement PTQ for APoT FakeQuant"


### Summary:
This PR implements PTQ for APoT FakeQuant. It runs models (Resnet-18 pre-trained model, ImageNet dataset) to compare accuracy metrics for different qconfig settings of uniform vs. APoT quantized activation and weight. 

According to the collected accuracy stats, model #2 (uniform activation and APoT weight) appears to have a slight improvement in accuracy compared to model #1 (uniform activation and uniform weight) for 8-bit and significant improvement for 4-bit (see "Accuracy Stats" section below).

### Test Plan:
Run models with: `python test/quantization/core/experimental/fx_graph_mode_apot.py`

### Accuracy Stats: 
8-bit (Uniform int8, APoT b = 8 k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.43% (Top-1), 85.62% (Top-5)

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.51% (Top-1), 85.78% (Top-5)

**Model #3:** APoT activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.32% (Top-1), 85.78% (Top-5)

4-bit (Uniform int4, APoT b = 4 k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 45.63% (Top-1), 71.96% (Top-5)

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.24% (Top-1), 85.56% (Top-5)

**Model #3:** APoT activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 45.40% (Top-1), 76.21% (Top-5)

**Full Precision model (FX Graph Mode quantized)**
Evaluation accuracy on test dataset: 69.76% (Top-1), 89.08% (Top-5)

**Eager mode quantized model**
Evaluation accuracy on test dataset: 69.49% (Top-1), 88.90% (Top-5)

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g
